File: plugins/stream-debugger-core/src/com/intellij/debugger/streams/trace/impl/interpret/ValuesOrderInfo.java
// Copyright 2000-2024 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
package com.intellij.debugger.streams.trace.impl.interpret;

import com.intellij.debugger.streams.trace.TraceElement;
import com.intellij.debugger.streams.trace.TraceInfo;
import com.intellij.debugger.streams.wrapper.StreamCall;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Collections;
import java.util.List;
import java.util.Map;

/**
 * @author Vitaliy.Bibaev
 */
public class ValuesOrderInfo implements TraceInfo {
  private final StreamCall myStreamCall;
  private final Map<Integer, TraceElement> myValuesOrderAfter;
  private final Map<Integer, TraceElement> myValuesOrderBefore;

  ValuesOrderInfo(@NotNull StreamCall call, @NotNull Map<Integer, TraceElement> before, @NotNull Map<Integer, TraceElement> after) {
    myStreamCall = call;
    myValuesOrderBefore = before;
    myValuesOrderAfter = after;
  }

  @Override
  public @NotNull StreamCall getCall() {
    return myStreamCall;
  }

  @Override
  public @NotNull Map<Integer, TraceElement> getValuesOrderBefore() {
    return myValuesOrderBefore;
  }

  @Override
  public @NotNull Map<Integer, TraceElement> getValuesOrderAfter() {
    return myValuesOrderAfter;
  }

  @Override
  public @Nullable Map<TraceElement, List<TraceElement>> getDirectTrace() {
    return null;
  }

  @Override
  public @Nullable Map<TraceElement, List<TraceElement>> getReverseTrace() {
    return null;
  }

  public static TraceInfo empty(@NotNull StreamCall call) {
    return new ValuesOrderInfo(call, Collections.emptyMap(), Collections.emptyMap());
  }
}

File: plugins/stream-debugger-core/src/com/intellij/debugger/streams/core/trace/impl/interpret/ValuesOrderInfo.kt
package com.intellij.debugger.streams.core.trace.impl.interpret

import com.intellij.debugger.streams.core.trace.TraceElement
import com.intellij.debugger.streams.core.trace.TraceInfo
import com.intellij.debugger.streams.core.wrapper.StreamCall

class ValuesOrderInfo(
  private val streamCall: StreamCall,
  private val before: Map<Int, TraceElement>,
  private val after: Map<Int, TraceElement>,
  private val direct: Map<TraceElement, List<TraceElement>>? = null,
  private val reverse: Map<TraceElement, List<TraceElement>>? = null,
) : TraceInfo {
  constructor(call: StreamCall, before: Map<Int, TraceElement>, after: Map<Int, TraceElement>) : this(call, before, after, null, null)

  override fun getCall(): StreamCall = streamCall

  override fun getValuesOrderBefore(): Map<Int, TraceElement> = before

  override fun getValuesOrderAfter(): Map<Int, TraceElement> = after

  override fun getDirectTrace(): Map<TraceElement, List<TraceElement>>? = direct

  override fun getReverseTrace(): Map<TraceElement, List<TraceElement>>? = reverse

  companion object {
    fun empty(call: StreamCall): TraceInfo {
      return ValuesOrderInfo(call, emptyMap(), emptyMap())
    }
  }
}